File: go/binary-search/binary_search.go
// Binary Search exercise.
package binarysearch

// Binary search for the given key in the list of ints.
// @param list: Slice of integers in sorted order.
// @param key: The number to search for.
// @returns: index where the key was found or -1 if not found.
func SearchInts(list []int, key int) int {
	// degenerate case, empty slice.
	if len(list) == 0 {
		return -1
	}
	var middle int
	var value int
	start := 0
	end := len(list) - 1

	for end >= start {
		// Degenerate case one space.
		if start == end {
			if list[start] != key {
				return -1 // Not found.
			}
			return start // Found
		}

		middle = start + ((end - start) / 2)
		value = list[middle]

		if value == key {
			return middle
		}

		if value > key {
			end = middle - 1
		}

		if value < key {
			start = middle + 1
		}
	}
	return -1
}

File: php/binary-search/BinarySearch.php
<?php
// Binary Search exercise.

declare(strict_types=1);

// Find a value in a sorted array using Binary Search
// Chose to do it iteratively this time.
// @param $needle: The value to search for
// @param $haystack: The values to search through.
// @returns: Index where the value was found or -1 if not found.
function find(int $needle, array $haystack): int
{
    if (count($haystack) == 0) {
        return -1;
    }

    $start = 0;
    $end = count($haystack) - 1;
    while (true) {
        if ($end - $start <= 0) {
            if ($haystack[$start] == $needle) {
                return $start;
            }
            return -1;
        }

        $middle = $start + intval(($end - $start) / 2);
        $value = $haystack[$middle];
        
        if ($value == $needle) {
            return $middle;
        } 
        if ($value > $needle) {
            $end = $middle - 1;
        }
        if ($value < $needle) {
            $start = $middle + 1;
        }
    }
}